### ISSUE TITLE
In-app update only works once, after that the same version is installed again and again

### ISSUE THREAD

**Issue by thhook** [microsoft/appcenter-sdk-android#1301]
### **Description**

The in-app update only works once. After that it keeps installing the same version over and over again without downloading further releases.

### **Repro Steps**

1. Download and install an app with enabled AppCenter distribution
2. Add a newer release of the app in AppCenter
3. Open the app on the device: an update dialog is shown
4. Download the new release and install it when finished
5. Add another newer release in AppCenter
6. Open the app again: an update dialog with the correct new release version is shown. Yet when clicking on "download" it always reuses the already installed current version, which leads to a kind of "endless loop"

### **Details**

1. Which SDK version are you using?
    - 2.4.1
2. Which OS version did you experience the issue on?
    - Android 9
3. What device version did you see this error on?  Were you using an emulator or a physical device?
    - Samsung Galaxy S8+ physical device
4. What third party libraries are you using?
 
5. Please enable verbose logging for your app using `AppCenter.setLogLevel(Log.VERBOSE)` before your call to `AppCenter.start(...)` and include the logs here:

    - First app session, no updates available
```
2019-10-31 15:02:24.099 5283-5283/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: InstallerPackageName=com.google.android.packageinstaller
2019-10-31 15:02:24.145 5283-5283/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: No token, need to open browser to url=https://install.appcenter.ms/apps/<omitted>/update-setup/?release_hash=c943efaf261c54447fd5281e53707efb1b58bed14e88141ec06606ed3c13e462&redirect_id=com.bmw.CNRDAppCenterTest&redirect_scheme=appcenter&request_id=a85c0e43-cd64-4b17-bd00-eff0dbb89836&platform=Android&enable_failure_redirect=true&install_id=605cf24e-9445-489c-8dbc-b378d5e67068
2019-10-31 15:02:24.157 5283-5283/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Default browser seems to be com.android.chrome/com.google.android.apps.chrome.IntentDispatcher
2019-10-31 15:02:24.157 5283-5283/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: And its not the picker.
2019-10-31 15:02:24.157 5283-5283/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Launch browser=com.android.chrome/com.google.android.apps.chrome.IntentDispatcher
2019-10-31 15:02:29.558 5283-5283/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: com.microsoft.appcenter.distribute.DeepLinkActivity.getIntent()=Intent { act=android.intent.action.VIEW cat=[android.intent.category.BROWSABLE] dat=appcenter://updates/ flg=0x14400000 pkg=com.bmw.CNRDAppCenterTest cmp=com.bmw.CNRDAppCenterTest/com.microsoft.appcenter.distribute.DeepLinkActivity (has extras) }
2019-10-31 15:02:29.558 5283-5283/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Intent requestId=a85c0e43-cd64-4b17-bd00-eff0dbb89836
2019-10-31 15:02:29.558 5283-5283/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Intent distributionGroupId=00000000-0000-0000-0000-000000000000
2019-10-31 15:02:29.558 5283-5283/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Intent updateToken passed=true
2019-10-31 15:02:29.558 5283-5283/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Intent updateSetupFailed passed=false
2019-10-31 15:02:29.558 5283-5283/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Intent testerAppUpdateSetupFailed passed=false
2019-10-31 15:02:29.579 5283-5283/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Stored redirection parameters.
2019-10-31 15:02:29.579 5283-5283/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: No sessions were logged before, ignore sending of the distribution start session log.
2019-10-31 15:02:29.580 5283-5283/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Get latest release details...
2019-10-31 15:02:29.580 5283-5283/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Check if we need to report release installation..
2019-10-31 15:02:29.580 5283-5283/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Current release was already reported, skip reporting.
2019-10-31 15:02:29.583 5283-5442/com.bmw.CNRDAppCenterTest V/AppCenterDistribute: Calling https://api.appcenter.ms/v0.1/sdk/apps/****************************f8a259eb/releases/latest?release_hash=c943efaf261c54447fd5281e53707efb1b58bed14e88141ec06606ed3c13e462...
2019-10-31 15:02:29.583 5283-5442/com.bmw.CNRDAppCenterTest V/AppCenterDistribute: Headers: {x-api-token=********************************3559000f}
2019-10-31 15:02:29.678 5283-5283/com.bmw.CNRDAppCenterTest V/AppCenterDistribute: Already checking or checked latest release.
2019-10-31 15:02:30.641 5283-5283/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Check if latest release is more recent.
2019-10-31 15:02:30.642 5283-5283/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Latest release more recent=false
```
    - First update available, installing it and open the app when done
```
2019-10-31 15:03:18.907 5283-5283/com.bmw.CNRDAppCenterTest I/AppCenterDistribute: Launcher activity restarted.
2019-10-31 15:03:18.946 5283-5283/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Get latest release details...
2019-10-31 15:03:18.946 5283-5283/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Check if we need to report release installation..
2019-10-31 15:03:18.947 5283-5283/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Current release was already reported, skip reporting.
2019-10-31 15:03:18.949 5283-5536/com.bmw.CNRDAppCenterTest V/AppCenterDistribute: Calling https://api.appcenter.ms/v0.1/sdk/apps/****************************f8a259eb/releases/latest?release_hash=c943efaf261c54447fd5281e53707efb1b58bed14e88141ec06606ed3c13e462...
2019-10-31 15:03:18.949 5283-5536/com.bmw.CNRDAppCenterTest V/AppCenterDistribute: Headers: {x-api-token=********************************3559000f}
2019-10-31 15:03:19.442 5283-5283/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Check if latest release is more recent.
2019-10-31 15:03:19.442 5283-5283/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Latest release more recent=true
2019-10-31 15:03:19.444 5283-5283/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Latest release is more recent.
2019-10-31 15:03:19.444 5283-5283/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Show default update dialog.
2019-10-31 15:03:23.929 5283-5283/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Show new unknown sources dialog.
2019-10-31 15:03:27.966 5283-5283/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Schedule download...
2019-10-31 15:03:27.967 5283-5283/com.bmw.CNRDAppCenterTest V/AppCenterDistribute: Already checking or checked latest release.
2019-10-31 15:03:27.969 5283-5569/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Start downloading new release from https://rink.hockeyapp.net/api/2/apps/<omitted>/app_versions/2?format=apk&avtoken=<omitted>&download_origin=hockeyapp&mctoken=<omitted>
2019-10-31 15:03:28.026 5283-5569/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Start download 2.0 (2) update.
2019-10-31 15:03:30.190 5283-5586/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Check download id=5276
2019-10-31 15:03:30.225 5283-5586/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Download was successful for id=5276
2019-10-31 15:03:30.236 5283-5586/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Download 2.0 (2) update completed.
2019-10-31 15:03:30.236 5283-5586/com.bmw.CNRDAppCenterTest I/AppCenterDistribute: Show install UI for content://downloads/all_downloads/5276
2019-10-31 15:03:30.268 5283-5586/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Stored release details: group id=00000000-0000-0000-0000-000000000000 release hash=f7db03fcfcc30537ce9718cca0a02f55219873432e76f995337f8219dcb95b2e release id=2
2019-10-31 15:03:48.089 5887-5887/? D/AppCenterDistribute: InstallerPackageName=com.google.android.packageinstaller
2019-10-31 15:03:48.105 5887-5887/? D/AppCenterDistribute: Get latest release details...
2019-10-31 15:03:48.105 5887-5887/? D/AppCenterDistribute: Check if we need to report release installation..
2019-10-31 15:03:48.106 5887-5887/? D/AppCenterDistribute: Current release was updated but not reported yet, reporting..
2019-10-31 15:03:48.113 5887-5949/? V/AppCenterDistribute: Calling https://api.appcenter.ms/v0.1/sdk/apps/****************************f8a259eb/releases/latest?release_hash=f7db03fcfcc30537ce9718cca0a02f55219873432e76f995337f8219dcb95b2e&distribution_group_id=00000000-0000-0000-0000-000000000000&downloaded_release_id=2...
2019-10-31 15:03:48.113 5887-5949/? V/AppCenterDistribute: Headers: {x-api-token=********************************3559000f}
2019-10-31 15:03:49.162 5887-5887/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Successfully reported app update for downloaded release hash (f7db03fcfcc30537ce9718cca0a02f55219873432e76f995337f8219dcb95b2e), removing from store..
2019-10-31 15:03:49.164 5887-5887/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Check if latest release is more recent.
2019-10-31 15:03:49.165 5887-5887/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Latest release more recent=false
```
    - Second update available. Note how the same download id (5276) is used again
```
2019-10-31 15:04:23.335 5887-5887/com.bmw.CNRDAppCenterTest I/AppCenterDistribute: Launcher activity restarted.
2019-10-31 15:04:23.375 5887-5887/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Get latest release details...
2019-10-31 15:04:23.375 5887-5887/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Check if we need to report release installation..
2019-10-31 15:04:23.375 5887-5887/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Current release was already reported, skip reporting.
2019-10-31 15:04:23.378 5887-6031/com.bmw.CNRDAppCenterTest V/AppCenterDistribute: Calling https://api.appcenter.ms/v0.1/sdk/apps/****************************f8a259eb/releases/latest?release_hash=f7db03fcfcc30537ce9718cca0a02f55219873432e76f995337f8219dcb95b2e...
2019-10-31 15:04:23.378 5887-6031/com.bmw.CNRDAppCenterTest V/AppCenterDistribute: Headers: {x-api-token=********************************3559000f}
2019-10-31 15:04:23.722 5887-5887/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Check if latest release is more recent.
2019-10-31 15:04:23.723 5887-5887/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Latest release more recent=true
2019-10-31 15:04:23.725 5887-5887/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Latest release is more recent.
2019-10-31 15:04:23.725 5887-5887/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Show default update dialog.
2019-10-31 15:04:26.835 5887-5887/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Schedule download...
2019-10-31 15:04:26.874 5887-6043/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Download was successful for id=5276
2019-10-31 15:04:26.893 5887-6043/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Download 2.0 (3) update completed.
2019-10-31 15:04:26.893 5887-6043/com.bmw.CNRDAppCenterTest I/AppCenterDistribute: Show install UI for content://downloads/all_downloads/5276
2019-10-31 15:04:26.923 5887-6043/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Stored release details: group id=00000000-0000-0000-0000-000000000000 release hash=5ce00fdee8722ac8aa6cc45b41c284bde02e3d69c7854a554225f4ecd37a1e34 release id=3
2019-10-31 15:04:48.992 6194-6194/? D/AppCenterDistribute: InstallerPackageName=com.google.android.packageinstaller
2019-10-31 15:04:49.008 6194-6194/? D/AppCenterDistribute: Get latest release details...
2019-10-31 15:04:49.009 6194-6194/? D/AppCenterDistribute: Check if we need to report release installation..
2019-10-31 15:04:49.009 6194-6194/? D/AppCenterDistribute: New release was downloaded but not installed yet, skip reporting.
2019-10-31 15:04:49.017 6194-6255/? V/AppCenterDistribute: Calling https://api.appcenter.ms/v0.1/sdk/apps/****************************f8a259eb/releases/latest?release_hash=f7db03fcfcc30537ce9718cca0a02f55219873432e76f995337f8219dcb95b2e...
2019-10-31 15:04:49.017 6194-6255/? V/AppCenterDistribute: Headers: {x-api-token=********************************3559000f}
2019-10-31 15:04:49.913 6194-6194/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Stored release hash doesn't match current installation, probably downloaded but not installed yet, keep in store
2019-10-31 15:04:49.913 6194-6194/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Check if latest release is more recent.
2019-10-31 15:04:49.914 6194-6194/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Latest release more recent=true
2019-10-31 15:04:49.922 6194-6194/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Latest release is more recent.
2019-10-31 15:04:49.924 6194-6194/com.bmw.CNRDAppCenterTest D/AppCenterDistribute: Show default update dialog.
```

It looks as if the relevant SharedPreferences key (PREFERENCE_KEY_DOWNLOAD_ID) as well as the DownloadManager's cache are not cleared after a successful update (as in `DownloadManagerReleaseDownloader.cancel()`).

**Comment by annakocheshkova** [comment]
Indeed. I can reproduce it. Thank you so much for the report!

**Comment by fabiendem** [comment]
Thanks for raising this so clearly.
I ended up seeing this issue/message by luck, before deploying an app to 100s of users and didn't really test again this flow with 2.4.1 (I did most of my testing back in time on 2.3.0).
Would be good if MS could communicate loudly about such critical issues 😕, like, before we deploy this and put ourselves in complicated situations.

**Comment by ahbilal-ms** [comment]
@thhook @fabiendem Thanks for raising this issue and apologies for all the inconvenience this is causing. You are absolutely right that we need to communicate these issues out loud and do that quickly. I will work with the team to improve our escalation and communication process. Also, while this issue was raised yesterday, our team has started looking into it and we will be releasing the update soon.

**Comment by ahbilal-ms** [comment]
@thhook @fabiendem, a quick update: The team has fixed the bug and are now testing it. We will be attempting to release the updated SDK by Monday. I will update you if there are any changes in the plan. Thanks

**Comment by guperrot** [comment]
One tester found an issue, if the download fails it also makes in-app updates stuck on that version.
We are investigating to see if it's a regression with the hotfix or an older bug before releasing.

On a side note, users currently on the app being stuck would not benefit from a fix as they are already stuck. To unstuck users, the updates need to be set as mandatory (it changes the way the cache behaves). It will still trigger the SDK 2.4.0/2.4.1 bug the first time (reinstalling the same version) but on restart with the reinstalled package it will clear the cache if the update was mandatory, then a second prompt will appear and this one will download the right package (the cache is not cleared if the updates are optional).

**Comment by guperrot** [comment]
The cancel download bug has also been fixed, the SDK is being prepared for release again.

**Comment by guperrot** [comment]
The SDK has just been released with the distribute fixes: version 2.5.0.
